### PR TITLE
Add nicer error when running out disk space

### LINKED ISSUE #3739: Nicer error for out of disk space when writing a FITS file
A user tried to write a file to disk that they didn't have space for and got an exception like this:

```
"/usr/stsci/ssbx/python/lib/python2.7/site-packages/astropy/io/fits/util.py
", line 653, in <lambda>
    write = lambda a, f: a.tofile(f)
IOError: 369098752 requested and 45004992 written
```

Of course, all this means is that they tried to write X bytes and only Y bytes were written where Y < X.  This could happen for a number of reasons, but lack of disk space is by far the most likely culprit.  I believe somewhere in `astropy.utils` there is a utility function to check space on a filesystem being written to.  I wouldn't call that in advance of writing just because in most cases it's unnecessary overhead.  But we could catch exceptions like the one above and _then_ run the disk space check to confirm that that's the culprit (and otherwise still raise an exception but note that the bytes could not be written for an unknown reason).

### PR DESCRIPTION
**Pull request by bsipocz**
Checking disk space before writing hdulist to file system. This is a rebased version of #4065 and should close #4065 and fix #3739.

~~Will need changelog and possible fixes.~~

### CODE REVIEW

**Comment by MSeifert04** [comment]
There is a one-off error in the github links. Should be #4065, right?

File: astropy/io/fits/hdu/hdulist.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import shutil
 import sys
 import warnings
+import numpy as np
```
Comment (MSeifert04) [inline astropy/io/fits/hdu/hdulist.py]: Not sure but I think this is not necessary.

File: astropy/io/fits/tests/test_core.py [hunk 2]
```diff
@@ -787,6 +790,7 @@ def test_updated_file_permissions(self):
 
         old_mode = os.stat(filename).st_mode
 
+        filename = self.temp('test.fits')
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_core.py]: why this change? This is already present in 786. rebase error?

**Comment by bsipocz** [comment]
@MSeifert04 - This is not yet ready for review, I'm still fixing the errors, but have though it will be useful to see what travis says on the rebase. Since the original PR wasn't touched in the last year I expect a few odd things due to the rebase.

**Comment by MSeifert04** [comment]
ok, I'll wait :-)

File: astropy/io/fits/hdu/hdulist.py [hunk 3]
```diff
@@ -986,20 +984,18 @@ def _flush_resize(self):
...
-
-            if sys.platform.startswith('win'):
-                # Collect a list of open mmaps to the data; this well be used
-                # later.  See below.
-                mmaps = [(idx, _get_array_mmap(hdu.data), hdu.data)
-                         for idx, hdu in enumerate(self) if hdu._has_data]
-
-            hdulist._file.close()
-            self._file.close()
+            with self.fromfile(new_file, mode='append') as hdulist:
```
Comment (bsipocz) [inline astropy/io/fits/hdu/hdulist.py]: While I haven't found an issue about this, there was a small bug here. This method left a few open files behind in the case of having an exception in the loop in the line below. Adding the with statement forced those temporary files to be closed.

**Comment by bsipocz** [comment]
Honestly the exact same thing passes locally and failing on travis, I'm at my wits end. Now I'm running some debugging on ``testing_branch``.  
I'm learning the lesson the hard way, don't do *trivial* small fixes to the fits module as it will certainly be a can of worms that eat up many hours to solve 😕

**Comment by MSeifert04** [comment]
I'll have a look - I have a hunch what could be causing this.

**Comment by bsipocz** [comment]
I've started the test locally on a linux box, too, I would guess the failure would be down to something different on osx.

**Comment by MSeifert04** [comment]
That's definitly a weird error. The idea I had didn't work out.

> I'm learning the lesson the hard way, don't do trivial small fixes to the fits module as it will certainly be a can of worms that eat up many hours to solve

Unfortunatly, yes.

**Comment by bsipocz** [comment]
I have a solution that works, but gave myself 30 more minutes to find a better one.

**Comment by MSeifert04** [comment]
Nice one, I was able to fix it by deleting `exc` at the end of the test (https://travis-ci.org/MSeifert04/astropy) I guess the problem is a reference cycle when catching the exception.

**Comment by MSeifert04** [comment]
oh, seems I was too slow, you got it working as well. 👍

**Comment by bsipocz** [comment]
@MSeifert04 - As we test the exc it shouldn't be deleted. Anyway, I've found a solution that both works and I think logical, ``hdulist.close()`` is the one that should take care of closing the files, no matter what happens in the meantime. So added a try/finally there, and now everything works as expected. 

This is ready for review, and being around for a long time, now hopefully will make it into 1.3.

File: astropy/io/fits/hdu/hdulist.py [hunk 4]
```diff
@@ -694,14 +694,13 @@ def writeto(self, fileobj, output_verify='exception', overwrite=False,
...
+        dirname = os.path.dirname(hdulist._file.name)
 
-        for hdu in self:
-            hdu._prewriteto(checksum=checksum)
-            try:
+        with _free_space_check(self, dirname=dirname):
+            for hdu in self:
+                hdu._prewriteto(checksum=checksum)
                 hdu._writeto(hdulist._file)
-            finally:
```
Comment (pllim) [inline astropy/io/fits/hdu/hdulist.py]: The removal of `try-except-finally` here assume that it only tries to catch "out of space" error. Is that always true? What if write fails for other reason? Then it won't clean up properly without the `finally`. Or perhaps you moved the `finally` to L731 below?

File: astropy/io/fits/hdu/hdulist.py [hunk 5]
```diff
@@ -724,16 +723,19 @@ def close(self, output_verify='exception', verbose=False, closed=True):
...
 
-        if self._file:
-            if self._file.mode in ['append', 'update']:
-                self.flush(output_verify=output_verify, verbose=verbose)
-
-            if closed and hasattr(self._file, 'close'):
-                self._file.close()
+        try:
+            if self._file:
+                if (self._file.mode in ['append', 'update']
```
Comment (pllim) [inline astropy/io/fits/hdu/hdulist.py]: The two `if`s can be combined into one? Same comment for L732 in `finally`.

File: astropy/io/fits/tests/test_core.py [hunk 6]
```diff
@@ -1015,6 +1017,63 @@ def test_filename_with_colon(self):
...
+    def test_writeto_full_disk(self, monkeypatch):
+        """
+        Test that it gives a readable error when trying to write an hdulist
+        to a full disk.
+        """
+        def _writeto(self, array):
+            raise OSError("Fake error raised when writing file.")
+
+        def get_free_space_in_dir(path):
+            return 0
```
Comment (pllim) [inline astropy/io/fits/tests/test_core.py]: What if the free space is `>0` but just not enough? That does not seem to be tested.

File: astropy/io/fits/util.py [hunk 7]
```diff
@@ -857,3 +859,22 @@ def _get_array_mmap(array):
...
+            hdulist = [hdulist, ]
+        if dirname is None:
+            dirname = os.path.dirname(hdulist._file.name)
+        if os.path.isdir(dirname):
+            free_space = data.get_free_space_in_dir(dirname)
+            hdulist_size = np.sum(hdu.size for hdu in hdulist)
+            for hdu in hdulist:
+                hdu._close()
+            if free_space < hdulist_size:
+                raise OSError("Not enough space on disk. " + str(exc))
```
Comment (pllim) [inline astropy/io/fits/util.py]: Since you already have the size info here (how much available vs how much needed), why not include them in the error message? I would find that useful personally.

File: astropy/io/fits/tests/test_core.py (hunk 6 above)
Comment (bsipocz) [inline astropy/io/fits/tests/test_core.py]: that should be fine, we just need to set a small enough number for mocking the test. I think.

File: astropy/io/fits/util.py (hunk 7 above)
Comment (bsipocz) [inline astropy/io/fits/util.py]: Sure, I'll add that.

File: astropy/io/fits/hdu/hdulist.py [hunk 8]
```diff
@@ -48,7 +48,7 @@ def fitsopen(name, mode='readonly', memmap=None, save_backup=False,
     save_backup : bool, optional
         If the file was opened in update or append mode, this ensures that a
         backup of the original file is saved before any changes are flushed.
-        The backup has the same name as the original file with ".bak" appended.
+        The backup has the same name as the original file  ".bak" appended.
```
Comment (bsipocz) [inline astropy/io/fits/hdu/hdulist.py]: I haven't noticed this, it came from the previous PR. Fixed.

File: CHANGES.rst [hunk 9]
```diff
@@ -55,10 +55,12 @@ New Features
   - Check for self-consistency of ECSV header column names. [#5463]
 
   - Produce warnings when writing an IPAC table from an astropy table that
-    contains metadata not supported by the IPAC format [#4700]
+    contains metadata not supported by the IPAC format. [#4700]
 
 - ``astropy.io.fits``
 
+  - Checking available disk space before writing out file. [#5500, #4065]
```
Comment (bsipocz) [inline CHANGES.rst]: Fixed

File: astropy/io/fits/util.py (hunk 7 above)
Comment (bsipocz) [inline astropy/io/fits/util.py]: However it the needed is sometimes misleading, as the exception can also come from the writing of a since hdu (thus it's not always the total size of the whole hdulist).

File: astropy/io/fits/hdu/hdulist.py [hunk 10]
```diff
@@ -694,14 +694,13 @@ def writeto(self, fileobj, output_verify='exception', overwrite=False,
         # but only if the file doesn't exist.
         fileobj = _File(fileobj, mode='ostream', overwrite=overwrite)
         hdulist = self.fromfile(fileobj)
+        dirname = os.path.dirname(hdulist._file.name)
```
Comment (bsipocz) [inline astropy/io/fits/hdu/hdulist.py]: I assumed that as the hdulist was just created from a a file, this should be fine. Tests pass. Would you prefer to have it in a try/except? I'm sure I'm not able to consider all possible corner case here.

File: astropy/io/fits/util.py (hunk 7 above)
Comment (pllim) [inline astropy/io/fits/util.py]: You are right, but reporting the numbers being compared is still useful, no?

File: astropy/io/fits/hdu/hdulist.py (hunk 10 above)
Comment (pllim) [inline astropy/io/fits/hdu/hdulist.py]: `try-except` here does not hurt, if nothing else, at least it would be consistent with your other change above.

File: astropy/io/fits/hdu/hdulist.py (hunk 5 above)
Comment (bsipocz) [inline astropy/io/fits/hdu/hdulist.py]: Done

File: astropy/io/fits/hdu/hdulist.py (hunk 4 above)
Comment (bsipocz) [inline astropy/io/fits/hdu/hdulist.py]: Yes, I think the try/finally in hdulist.close() should be able to deal with all kinds of interrupted write, given that the file was opened with ``with``, or the close() was called manually. 

As for the individual hdu's, the exception handling in ``_free_space_check()`` now calls hdu._close() for all of them.

File: astropy/io/fits/hdu/hdulist.py (hunk 10 above)
Comment (bsipocz) [inline astropy/io/fits/hdu/hdulist.py]: OK, I've added it back. However not sure whether that case would make any sense here, if dirname is ``None``, it will try to get the directory name from ``self._file.name``, which may not exist either. Anyway, tests pass both ways, so any problem will be an extreme corner case that only comes up when there isn't enough space left on disk :)

**Comment by bsipocz** [comment]
@pllim - Thanks for the comments. I'm not sure whether it will work on quota systems, but that really is an issue for ``astropy.utils.data.get_free_space_in_dir()``. If it can handle available space from the quota, then this works, it it doesn't, then the original exception message will come back.

**Comment by pllim** [comment]
LGTM :+1:

**Comment by eteq** [comment]
Does not seem to have reached consensus (and needs a rebase) and well past feature freeze, so pushing to 2.0 milestone.

**Comment by bsipocz** [comment]
@eteq - I assumed the thumbs ups from @pllim mean that this is good to go, though it's fine to get this into 2.0 once I rebase. I'll also move the changelog entry to 2.0.

File: astropy/io/fits/hdu/base.py [hunk 11]
```diff
@@ -651,69 +650,75 @@ def _writedata_direct_copy(self, fileobj):
...
-        self._data_offset = datloc
-        self._data_size = datsize
-        self._data_replaced = False
+        try:
+            dirname = os.path.dirname(fileobj._file.name)
+        except AttributeError:
+            dirname = None
+
+        with _free_space_check(self, dirname):
+            # For now fileobj is assumed to be a _File object
```
Comment (embray) [inline astropy/io/fits/hdu/base.py]: This is a lot of stuff to shove under a single with-statement.  My rule of thumb is if the entire block can't fit on my screen it should probably be broken up somehow.  In this case I would take all of this and move it a `_writeto_internal` method, and then just have:

```
with _free_space_check(self, dirname):
    self._writeto_internal(...)
```

This is entirely subjective though so it's up to you.

Comment (bsipocz) [inline astropy/io/fits/hdu/base.py]: Thanks @embray this makes sense, I always learn from your rule of thumbs.

**Comment by bsipocz** [comment]
@pllim @eteq - Can we merge this soonish? The original issue (#3739) is labelled as ``package-novice`` thus will probably draw the attention of incoming GSoC applicants.

File: CHANGES.rst [hunk 12]
```diff
@@ -18,6 +18,8 @@ New Features
...
```
Comment (pllim) [inline CHANGES.rst]: Oh noes! I thought you fixed this but maybe it was lost in rebase. Can you please submit a small PR to fix this change log? If not, maybe someone with write priviledge to master can just change it there?

Comment (pllim) [inline CHANGES.rst]: Ok, nevermind. Turns out I can edit it on master. It is fixed.

Comment (bsipocz) [inline CHANGES.rst]: Surprising as I remember fixing it and GH also hid your comment about it.

Comment (pllim) [inline CHANGES.rst]: Astrobot to the rescue!
